api/src/projects/groups.ts

```diff
@@ -0,0 +1,51 @@
+import { createGroup, IGroup } from '@esri/arcgis-rest-groups';
+import { IAuthenticationManager } from '@esri/arcgis-rest-request';
+import 'isomorphic-fetch';
+import 'isomorphic-form-data';
+
+class Auth implements IAuthenticationManager {
+  public  portal: string
+  private token: string
+  
+  constructor(token: string) {
+    this.token = token.substring(7);
+  }
+
+  public getToken(): Promise<string> {
+    return new Promise((resolve, reject) => {
+      resolve(this.token);
+    });
+  }
+}
+
+const groupNames: {[key: string]: string} = {
+  project_managers: 'Project managers',
+  field_supervisors: 'Field supervisors',
+  data_collectors: 'Data Collectors',
+  viewers: 'Viewers'
+}
+  
+export async function createGroups(groups: string[], projectName: string, slug: string, owner: string, token: string) {
+  let i = 0;
+  const len = groups.length;
+  for (i; i < len; i++) {
+    const groupName: string = groupNames[groups[i]];
+    const group: IGroup = {
+      title: `${projectName}: ${groupName}`,
+      owner,
+      tags: [`project:${slug}`],
+      access: 'private'
+    }
+
+    const authentication: IAuthenticationManager = new Auth(token);
+    try {
+      await createGroup({
+        authentication,
+        group,
+        portal: process.env.ARCGIS_PORTAL_URL + '/rest',
+      });
+    } catch (e) {
+      throw new Error(JSON.stringify(e.response));
+    }
+  }
+}
```

Comment:
I question the design of this function for the following reasons:

1. We're creating groups synchronously despite the fact that they are not dependent on eachother. This is nicer on our Portal server, but will really extend the response time for this API endpoint we're creating. Instead, I suggest we just make all the group creation requests together.
2. We don't return anything from this function. We don't return any information about the groups that we create.  We could assume that if no errors are created, all groups are created; however this doesn't really feel RESTful.  Instead, I suggest that we return all resources created in the response body of the lambda creation endpoint.

Thinking about how we can meet these goals, I put together the following example utilizing (`Promise.all`)[]:

```js
interface ParallelOpResult {
  success: {[key: string]: any}[],
  failure: ArcGISRequestError[]
}

const createGroups = (
  groups: string[],
  projectName: string,
  slug: string,
  owner: string,
  token: string
): Promise<ParallelOpResult> => (
  Promise.all(groups
    // Create ArcGIS request input
    .map(
      (group: string): IGroup => ({
        title: `${projectName}: ${groupNames[group]}`,
        owner,
        tags: [`project:${slug}`],
        access: 'private'
      })
    )
    // Create group-creation promises
    .map(
      (group: IGroup) => createGroup({
        group,
        authentication: new Auth(token),
        portal: process.env.ARCGIS_PORTAL_URL + '/rest'
      })
      // Catch errors, return as error object
      .catch((err: ArcGISRequestError) => ({ err, group }))
    )
  )
  // Combine array of successes and failures into single object
  .then(
    (resps): ParallelOpResult =>
    resps.reduce(
      (output: ParallelOpResult, result) => ({
        success: output.success.concat(!result.err ? [result] : []),
        failure: output.failure.concat(result.err ? [result] : [])
      }),
      { success: [], failure: [] }
    )
  )
);
```

This will produce an object of `success` and `failure` arrays.  This feels like the first step to then allowing us to retry failures or deleting successes if there are any failures.